### DISCUSSION TITLE
Controller events

### DISCUSSION THREAD

**Discussion by Osmosis311** [uhppoted/uhppoted#7] [General] [answered]
Not an issue per se, just a general question if you have knowledge of the UHPPOTED boards: if I wanted to write my own controller on a Pi let's say, and just use the board for the reader connections and the relays, does talking to the board via API allow you to receive events in real-time when a card is presented to a reader?

**Comment by uhppoted** [comment]
_(converted issue to discussion on the basis that it's come up once or twice before so might be of general interest)_

**Comment by uhppoted** [comment, marked as answer]
Yes, the controllers do support sending events in real-time to a designated '_listen_' host. The core APIs to configure event listening are in [uhppote-core](https://github.com/uhppoted/uhppote-core):

- [get-listener](https://github.com/uhppoted/uhppote-core#getlistener)
- [set-listener](https://github.com/uhppoted/uhppote-core#setlistener)
- [record-special-events](https://github.com/uhppoted/uhppote-core#recordspecialevents)

Once you have configured a controller to send events to a host it will send a UDP event message on every event so the host has to have a _listening_ UDP connection established to receive the messages. Be aware that the default controller configuration does not send _door open_ and _close_ events so if you want those (most people seem to) you need to enable them with _record-special-events_.

The [CLI](https://github.com/uhppoted/uhppote-cli) supports all of the above:

- [get-listener](https://github.com/uhppoted/uhppote-cli#get-listener)
- [set-listener](https://github.com/uhppoted/uhppote-cli#set-listener)
- [record-special-events](https://github.com/uhppoted/uhppote-cli#record-special-events)
- [listen](https://github.com/uhppoted/uhppote-cli#listen)

so if you want to experiment before diving in and writing your own that's a good place to start. It is possible to wrap the _listen_ command in a _systemd_ service/daemon script so that it runs in the background, but be warned it hasn't been tested for long-term use and may not be robust enough in its current incarnation.

I haven't actually measured the delay between an e.g. a door open/close event and the corresponding UDP message but it appears to be sub-second'ish. Also ... please bear in mind that UDP isn't a reliable transport so messages can be expected to go missing entirely - although (_having diligently issued the standard disclaimer_) I've generally found UDP on a LAN connection to be "good enough" for practical purposes e.g. a card may have to be occasionally swiped twice or some such but that isn't that an unusual occurrence anyway.

**Reply by Osmosis311** [reply]
Thank you for the fast reply!!!!

I don't think we need the door open/close events.  I just don't want to use the controller for any access control logic.  My plan (hope) is that when someone badges at the reader, I can get an event with the decoded ID to my NodeJS app, and then do a lookup, etc, and then tell the controller to unlock that door.

How do you configure the controller to send events to a host?

Do you by any chance have an example event when a badge is read?  Does it include door ID + the card ID?

Thanks!!!

**Reply by uhppoted** [reply]
The [uhppoted-nodejs](https://github.com/uhppoted/uhppoted-nodejs) NodeJS implementation of the API includes the functions mentioned above and there are examples of how to invoke the functions along with the responses at these links:

- [get-listener](https://github.com/uhppoted/uhppoted-nodejs#getlistener)
- [set-listener](https://github.com/uhppoted/uhppoted-nodejs#getlistener)
- [listen](https://github.com/uhppoted/uhppoted-nodejs#listen)
- [open-door](https://github.com/uhppoted/uhppoted-nodejs#opendoor)

The _listen_ link includes a sample event, and yup, it includes the door and card number. The project also includes a set of [examples](https://github.com/uhppoted/uhppoted-nodejs/tree/master/examples) to get you started.

Configuring a controller to send events to your app is straightforward - invoke _set-listener_ to set the IP address and port the app is listening on and that's about it.

The above should be enough if you're not intending to use the controller as a fallback for when e.g. the app is not running. If you do want to use the controller as a fallback there is an additional API (_set-host-control_) which hasn't been implemented as yet but which I'm happy to do (as long you're not in a rush).

**Comment by uhppoted** [comment]
BTW, there is apparently a way to let the operating system handle incoming network messages and dispatch them to a designated handler which may be a more robust way to implement a long-running event listener. 

It's on a TODO list somewhere, so if you do decide to use the CLI _listen_ command in a service wrapper, please let me know and I'll dig it out and bump up the priority.

**Comment by uhppoted** [comment]
From e-mail discussion:

I was just about to ask you to double check your IP configuration :-) .. great!  You need to bind your UDP server to the port the controller is sending events to i.e. if you have done: 

set-listener 405419896 192.168.1.100:60001

then your UDP server must be bound to 60001 to receive messages from the controller - it's just how UDP works.

On 2022-03-11 17:47, Danny Ackerman wrote:

> Found the issue! For some reason, no matter what IP address I used for the binding configuration (e.g. my computer's local IP or 0.0.0.0) nothing was getting received. I change the bind code in driver.js to sock.bind(6001); and it worked!!! 
 

Links:
------
[1]
https://github.com/uhppoted/uhppoted/discussions/7#discussioncomment-2340082
[2]
https://github.com/notifications/unsubscribe-auth/AOVVQCRX4O6JZQD6QUT5AQTU7NTIVANCNFSM5QHYMRXQ
[3]
https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&amp;mt=8&amp;pt=524675
[4]
https://play.google.com/store/apps/details?id=com.github.android&amp;referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub
